### PR TITLE
Request sender improvements

### PR DESCRIPTION
**Pull request by dejan2k**

### CODE REVIEW

File: Leanplum-SDK/Classes/Internal/Leanplum.m [hunk 1]
```diff
@@ -1110,22 +1111,12 @@ + (void)startWithUserId:(NSString *)userId
                     BOOL exitOnSuspend = [[[[NSBundle mainBundle] infoDictionary]
                         objectForKey:@"UIApplicationExitsOnSuspend"] boolValue];
                     LPRequest *request = [LPRequestFactory stopWithParams:nil];
-                    [[LPRequestSender sharedInstance] sendIfConnected:request sync:exitOnSuspend];
+                    request.sendSynchronously = exitOnSuspend;
```
Comment (milos1290) [inline Leanplum-SDK/Classes/Internal/Leanplum.m:1114]: Get rid of `sendSyncronously` flag

File: Leanplum-SDK/Classes/Managers/LPAppIconManager.m [hunk 2]
```diff
@@ -72,7 +72,9 @@ + (void)uploadAppIconsOnDevMode
     [request onError:^(NSError *error) {
         LPLog(LPError, @"Fail to upload app icons: %@", error.localizedDescription);
     }];
-    [[LPRequestSender sharedInstance] sendNow:request withDatas:requestDatas];
+    request.datas = requestDatas;
+    request.requestType = Immediate;
```
Comment (milos1290) [inline Leanplum-SDK/Classes/Managers/LPAppIconManager.m:76]: Move `requestType` to be as part of the builder pattern instead setting it like this

File: Leanplum-SDK/Classes/Managers/Networking/LPRequest.h [hunk 3]
```diff
@@ -27,20 +27,28 @@
...
 @interface LPRequest : NSObject
 
-@property (nonatomic, strong) NSString *apiMethod;
-@property (nonatomic, strong, nullable) NSDictionary *params;
 @property (atomic) BOOL sent;
 @property (nonatomic, copy, nullable) LPNetworkResponseBlock responseBlock;
 @property (nonatomic, copy, nullable) LPNetworkErrorBlock errorBlock;
 @property (nonatomic, strong) NSString *requestId;
+@property (nonatomic, assign) LPRequestType requestType;
+@property (nonatomic, assign) BOOL sendSynchronously;
```
Comment (milos1290) [inline Leanplum-SDK/Classes/Managers/Networking/LPRequest.h:42]: Get rid of `sendSynchronously`

File: Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m [hunk 4]
```diff
@@ -66,53 +65,67 @@ - (id)init
...
+                                               userInfo:@{NSLocalizedDescriptionKey: @"Device is offline"}]);
+        }
+
+        return false;
     }
+    
+    return true;
+}
+
+- (void)sendNow:(LPRequest *)request sync:(BOOL)sync
```
Comment (milos1290) [inline Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m:119]: Please get rid off `sync` flag from all methods and always execute on `LPOperationQueue`

File: Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m [hunk 5]
```diff
@@ -148,81 +161,15 @@ - (void)sendEventually:(LPRequest *)request sync:(BOOL)sync
...
-    }
-    LP_END_TRY
-}
-
-- (void)sendNow:(LPRequest *)request withData:(NSData *)data forKey:(NSString *)key
-{
-    [self sendNow:request withDatas:@{key: data}];
-    
-    [self.countAggregator incrementCount:@"send_now_with_data_lp"];
+    [self sendNow:request sync:request.sendSynchronously];
```
Comment (milos1290) [inline Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m:166]: Please get rid off `sync` flag from all methods and always execute on `LPOperationQueue`

**Review by milos1290 — CHANGES_REQUESTED** [review]

File: Leanplum-SDK/Classes/Managers/Networking/LPNetworkEngine.h [hunk 6]
```diff
@@ -33,4 +33,10 @@
 @property (nonatomic, strong) NSURLSessionConfiguration *sessionConfiguration;
 @property (strong, nonatomic) Leanplum_Reachability *reachability;
 
+/**
+ * Create Request Headers for network call
+ */
++ (NSDictionary *)createHeaders;
++ (NSDictionary *)notificationSettingsToRequestParams:(NSDictionary *)settings;
```
[the remark this answers is not in the document]
Comment (nzagorchev) [reply Leanplum-SDK/Classes/Managers/Networking/LPNetworkEngine.h:40]: It is only for the way those are sent in a request, I believe it should be here

File: Leanplum-SDK/Classes/Managers/LPAppIconManager.m (hunk 2 above)
Comment (nzagorchev) [reply Leanplum-SDK/Classes/Managers/LPAppIconManager.m:76]: I agree, it will make it more clear if the request will be batched or send immediately.

**Review by nzagorchev — APPROVED** [review]

**Review by milos1290 — APPROVED** [review]

**Pull request merged**